homeassistant/components/ollama_conversation/__init__.py

```diff
@@ -0,0 +1,263 @@
+    MAX_HISTORY_SECONDS,
+)
+from .models import ExposedEntity, MessageHistory, MessageRole
+
+_LOGGER = logging.getLogger(__name__)
+
+__all__ = [
+    "CONF_URL",
+    "CONF_PROMPT",
+    "CONF_MODEL",
+    "CONF_MAX_HISTORY",
+    "MAX_HISTORY_NO_LIMIT",
+    "DOMAIN",
+]
+
+CONFIG_SCHEMA = cv.config_entry_only_config_schema(DOMAIN)
+
+
+async def async_setup_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
+    """Set up Ollama conversation from a config entry."""
+    settings = {**entry.data, **entry.options}
+    client = ollama.AsyncClient(host=settings[CONF_URL])
+    try:
+        async with asyncio.timeout(DEFAULT_TIMEOUT):
+            await client.list()
+    except (TimeoutError, httpx.ConnectError) as err:
+        raise ConfigEntryNotReady(err) from err
+
+    hass.data.setdefault(DOMAIN, {})[entry.entry_id] = client
+
+    conversation.async_set_agent(hass, entry, OllamaAgent(hass, entry))
+    return True
+
+
+async def async_unload_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
+    """Unload Ollama conversation."""
+    hass.data[DOMAIN].pop(entry.entry_id)
+    conversation.async_unset_agent(hass, entry)
+    return True
+
+
+class OllamaAgent(conversation.AbstractConversationAgent):
+    """Ollama conversation agent."""
+
+    def __init__(self, hass: HomeAssistant, entry: ConfigEntry) -> None:
+        """Initialize the agent."""
+        self.hass = hass
+        self.entry = entry
+
+        # conversation id -> message history
+        self._history: dict[str, MessageHistory] = {}
+
+    @property
+    def supported_languages(self) -> list[str] | Literal["*"]:
+        """Return a list of supported languages."""
+        return MATCH_ALL
+
+    async def async_process(
+        self, user_input: conversation.ConversationInput
+    ) -> conversation.ConversationResult:
+        """Process a sentence."""
+        settings = {**self.entry.data, **self.entry.options}
+
+        client = self.hass.data[DOMAIN][self.entry.entry_id]
+        conversation_id = user_input.conversation_id or ulid.ulid_now()
+        model = settings[CONF_MODEL]
+
+        # Look up message history
+        message_history: MessageHistory | None = None
+        message_history = self._history.get(conversation_id)
+        if message_history is None:
+            # New history
+            #
+            # Render prompt and error out early if there's a problem
+            raw_prompt = settings[CONF_PROMPT]
+            try:
+                prompt = self._generate_prompt(raw_prompt)
+                _LOGGER.debug("Prompt: %s", prompt)
+            except TemplateError as err:
+                _LOGGER.error("Error rendering prompt: %s", err)
+                intent_response = intent.IntentResponse(language=user_input.language)
+                intent_response.async_set_error(
+                    intent.IntentResponseErrorCode.UNKNOWN,
+                    f"Sorry, I had a problem generating my prompt: {err}",
```

Comment:
Not for this PR, but it should be interesting to explore if we can have an error message that is for humans, and an error message with tech details.

Do you want your voice assistant to say: "Cannot concatenate None + Str" ?

Reply:
My Google TV brilliantly tells me "Something's not right" and then continues to work just fine :robot: